Add Epic Games Store integration

Quentame opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds 2 Epic Games Store calendars, for free and discount games.

It squashes 62 commits from Hacktoberfest 2022 old PR #81167, and took care of @joostlek, @nachonam and @epenet reviews.

I may eventually add price tracking for specific games in the future.

Screenshot:

<img width="1397" alt="Capture d’écran 2023-10-04 à 20 14 06" src="https://github.com/home-assistant/core/assets/14821482/710118f1-2e04-40d7-8d87-484d11f4ae94">


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to brand pull request: https://github.com/home-assistant/brands/pull/3820
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24798

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Quentame committed:
Add Epic Games Store integration

Squashed commit of the following PR: #81167

Quentame committed:
Bump epicstore-api to 0.1.7 as it handle better error 1004

Thanks to https://github.com/SD4RK/epicstore_api/commit/d7469f7c99508c06b3867fecbcf291ebf86c4c72

Quentame committed:
Use extra_state_attributes instead of overriding state_attributes

Quentame committed:
Review: change how config_flow.validate_input is handled

Quentame committed:
Use LanguageSelector and rename locale to language

Quentame commented:
@bramkragten: The new two-pane view from 2023.11 caused long label entity names not to be distinctive, not even a title or tooltip on mouse hover is present

<img width="1418" alt="image" src="https://github.com/home-assistant/core/assets/14821482/a17efd64-8eb8-4eb2-a279-664b111e715d">

_One is a discount calendar, one is a free games calendar (mystery)_ 🔮

Compare to before 2023.11 in the PR description.

A possible solution could be using a shorter name as the integration page ?
<img width="694" alt="image" src="https://github.com/home-assistant/core/assets/14821482/20f4f8e9-f6be-4c74-879d-4bef317bbefe">

Also a solution, on the two-pane calendar page, open the more info dialog when clicking ont the entity icon ?

Thanks

joostlek commented on homeassistant/components/epic_games_store/__init__.py:
```diff
@@ -0,0 +1,36 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Epic Games Store from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    coordinator = EGSCalendarUpdateCoordinator(hass, entry)
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
```suggestion
    coordinator = EGSCalendarUpdateCoordinator(hass, entry)
    await coordinator.async_config_entry_first_refresh()

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

joostlek commented on homeassistant/components/epic_games_store/calendar.py:
```diff
@@ -0,0 +1,110 @@
+        EGSCalendar(coordinator, entry.entry_id, CalendarType.FREE),
+        EGSCalendar(coordinator, entry.entry_id, CalendarType.DISCOUNT),
+    ]
+    async_add_entities(entities, True)
```
```suggestion
    async_add_entities(entities)
```
You're using a coordinator so data is already present when you add entities, no need to do another update

joostlek commented on homeassistant/components/epic_games_store/calendar.py:
```diff
@@ -0,0 +1,110 @@
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return the state attributes."""
+        return {
+            "games": self.coordinator.data[self._cal_type],
```
What does this return? We want to avoid putting a lot of data in these state attributes as they can blow up the state machine

Quentame replied:
`TypeError` or `KeyError`, more a Type as on test `tests/components/epic_games_store/test_config_flow.py::test_form_cannot_connect_wrong_param`

See

```shell
2023-12-28 11:01:23.456 WARNING  MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:48 [{'message': 'CatalogQuery/searchStore: Request failed with status code 400', 'locations': [{'line': 18, 'column': 9}], 'correlationId': 'e10ad58e-a4f9-4097-af5d-cafdbe0d8bbd', 'serviceResponse': '{"errorCode":"errors.com.epicgames.catalog.invalid_country_code","errorMessage":"Sorry the value you entered: en-US, does not appear to be a valid ISO country code.","messageVars":["en-US"],"numericErrorCode":5222,"originatingService":"com.epicgames.catalog.public","intent":"prod","errorStatus":400}', 'stack': None, 'path': ['Catalog', 'searchStore']}]
2023-12-28 11:01:23.456 ERROR    MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:82 Unexpected exception
Traceback (most recent call last):
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 78, in async_step_user
    await validate_input(self.hass, user_input)
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 51, in validate_input
    data["data"]["Catalog"]["searchStore"]["elements"]
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

But this case should not happen now as of new version of lib, so I refactored it in c0a3429e6063c38f7f67d0d56159df7f5077b2fd

joostlek replied:
Yes we can! Since this PR can't make 2024.2, I think we can make some nice services in a follow up to have it ready with 2024.3

Quentame replied:
Fixed in 73c20f34803b0a0ec242bf0740494f17a68f6f59

Quentame replied:
Fixed in f3d475385aa2fb8e6c994237e000270e4ab095f8

joostlek commented on homeassistant/components/epic_games_store/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+
+    try:
+        data["data"]["Catalog"]["searchStore"]["elements"]
+    except Exception:  # pylint: disable=broad-except
```
I think what we are ultimately catching is a KeyError right?

Quentame replied:
`TypeError` or `KeyError`, more a Type as on test `tests/components/epic_games_store/test_config_flow.py::test_form_cannot_connect_wrong_param`

See

```shell
2023-12-28 11:01:23.456 WARNING  MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:48 [{'message': 'CatalogQuery/searchStore: Request failed with status code 400', 'locations': [{'line': 18, 'column': 9}], 'correlationId': 'e10ad58e-a4f9-4097-af5d-cafdbe0d8bbd', 'serviceResponse': '{"errorCode":"errors.com.epicgames.catalog.invalid_country_code","errorMessage":"Sorry the value you entered: en-US, does not appear to be a valid ISO country code.","messageVars":["en-US"],"numericErrorCode":5222,"originatingService":"com.epicgames.catalog.public","intent":"prod","errorStatus":400}', 'stack': None, 'path': ['Catalog', 'searchStore']}]
2023-12-28 11:01:23.456 ERROR    MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:82 Unexpected exception
Traceback (most recent call last):
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 78, in async_step_user
    await validate_input(self.hass, user_input)
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 51, in validate_input
    data["data"]["Catalog"]["searchStore"]["elements"]
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

But this case should not happen now as of new version of lib, so I refactored it in c0a3429e6063c38f7f67d0d56159df7f5077b2fd

Quentame replied:
fixed in 14ca00a3c56cea41965ac676943957d035a86f37

Quentame replied:
Apply fix again in 5eda545a4695ad9956abe8a89e88b5e5ecb05200

Quentame committed:
Review: init-better use of hass.data.setdefault

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

Quentame committed:
Review: don't need to update at init

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

Quentame committed:
Revert "Review: don't need to update at init" --> not working otherwise

This reverts commit 1445a87c8e9b7247f1c9835bf2e2d7297dd02586.

Quentame committed:
Review: fix config_flow.validate_input/retactor following lib bump

joostlek commented on homeassistant/components/epic_games_store/calendar.py:
```diff
@@ -0,0 +1,110 @@
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        return self._event
```
```suggestion
        if event := self.coordinator.data[self._cal_type]:
            return _get_calendar_event(event[0])
        return None
```

joostlek commented on homeassistant/components/epic_games_store/calendar.py:
```diff
@@ -0,0 +1,110 @@
+            manufacturer="Epic Games Store",
+            name="Epic Games Store",
+        )
+        self._event: CalendarEvent | None = None
```
```suggestion
```

joostlek commented on homeassistant/components/epic_games_store/calendar.py:
```diff
@@ -0,0 +1,110 @@
+            self.coordinator.data[self._cal_type],
+        )
+        return [_get_calendar_event(event) for event in events]
+
+    async def async_update(self) -> None:
+        """Update entity state with the next upcoming event."""
+        event: list[dict[str, Any]] = self.coordinator.data[self._cal_type]
+        if event:
+            self._event = _get_calendar_event(event[0])
```
```suggestion
```

bramkragten commented:
> @bramkragten: The new two-pane view from 2023.11 caused long label entity names not to be distinctive, not even a title or tooltip on mouse hover is present

I guess a tooltip/hover makes sense there yeah

Quentame committed:
review: merge async_update function with event property

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

Quentame committed:
hassfest

Quentame committed:
Fix duplicates data from applied comment review 5035055

Quentame committed:
review: thanks to 5035055 async_add_entities update_before_add param is not required anymore

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

Quentame committed:
Fix Christmas special "Holiday sale" case

Quentame committed:
gen_requirements_all

joostlek commented on homeassistant/components/epic_games_store/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
```
unused

Quentame replied:
Fixed in 040cf945bb5346b6d42b3782b5061a13fb7b1f6b

Quentame commented on homeassistant/components/epic_games_store/__init__.py:
```diff
@@ -0,0 +1,68 @@
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_DATA,
```
Maybe an other name than "Data" ?

"Get free games data" ?

Quentame commented on homeassistant/components/epic_games_store/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Services
+    def async_data(call: ServiceCall) -> ServiceResponse:
+        """Handle data service call."""
+        return {
+            **__serialize_data(coordinator, "discount"),
+            **__serialize_data(coordinator, "free"),
+        }
```
Add a "type" param ?

Possible values could be "all", "free", "discount". Optional. Defaults to "all"

Return a list of games.

Quentame commented on homeassistant/components/epic_games_store/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+        hass.services.async_remove(DOMAIN, SERVICE_DATA)
```
Remove only if no entries left in the domain.

Quentame commented on homeassistant/components/epic_games_store/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    return unload_ok
+
+
+def __serialize_data(
+    coordinator: EGSCalendarUpdateCoordinator, cal_type: str
+) -> JsonObjectType:
+    """Serialize data to service response."""
+    return {
+        cal_type: [
+            {str(key): str(value) for key, value in games.items()}
+            for games in coordinator.data[cal_type]
+        ]
+    }
```
A shame that we need that for mypy

autinerd commented on homeassistant/components/epic_games_store/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.selector import LanguageSelector, LanguageSelectorConfig
+
+from .const import CONF_LANGUAGE, CONF_SUPPORTED_LANGUAGES, DOMAIN
```
`CONF_LANGUAGE` is available in `homeassistant.const`.

Quentame replied:
Done in 724ce255805e57ea05a7040f6e19d2b945ff67c7

autinerd commented on homeassistant/components/epic_games_store/const.py:
```diff
@@ -0,0 +1,24 @@
+
+DOMAIN = "epic_games_store"
+
+CONF_LANGUAGE = "language"
```
```suggestion
```

Quentame replied:
Done in 724ce255805e57ea05a7040f6e19d2b945ff67c7

autinerd commented on homeassistant/components/epic_games_store/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "codeowners": ["@Quentame"],
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/epic_games_store",
+  "iot_class": "cloud_polling",
```
```suggestion
  "integration_type": "service",
  "iot_class": "cloud_polling",
```

emontnemery commented on homeassistant/components/epic_games_store/helper.py:
```diff
@@ -0,0 +1,101 @@
+
+def get_country_from_language(language: str) -> str:
+    """Get the country code from language."""
+    excepts = {"ja": "JP", "ko": "KR", "zh-Hant": "CN"}
```
I don't think `"zh-Hant": "CN"` is correct, traditional Chinese writing is not used in mainland China. It is however used in Hong Kong, Singapore and Shanghai.

I'd suggest to double check where the epic game store is available, and add options for that to the config flow instead of trying to guess it here.

Quentame replied:
Fixed with 7a53234, the only thing the country change is the currency & price.

Tested with weird combinaisons, and working fine.
I put the selector is the config flow so Chinese people living in France (or anywhere else 😄) could get the local currency & price, with his native language 😉

Quentame committed:
Use CONF_LANGUAGE from HA const

Quentame committed:
Move CalendarType to const

Quentame committed:
manifest: integration_type -> service

Co-authored-by: Sid <27780930+autinerd@users.noreply.github.com>

Quentame committed:
calendar: remove date start/end assert

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Quentame committed:
const: rename SUPPORTED_LANGUAGES

Quentame committed:
hassfest

Quentame committed:
config: Move to ConfigFlowResult

Quentame commented on homeassistant/components/epic_games_store/const.py:
```diff
@@ -0,0 +1,24 @@
+
+CONF_LANGUAGE = "language"
+
+CONF_SUPPORTED_LANGUAGES = [
```
Done in 8b1d9ba

Quentame committed:
coordinator: main file comment

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Quentame commented on homeassistant/components/epic_games_store/coordinator.py:
```diff
@@ -0,0 +1,87 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class CalendarType(StrEnum):
```
Done in 10d0083

Quentame committed:
ruff & hassfest

Quentame committed:
review: do not guess country

Quentame committed:
Add @hacf-fr as codeowner

Quentame committed:
review: remove games extra_attrs
Was dropped somehow:
- 73c20f34803b0a0ec242bf0740494f17a68f6f59 review: move games extra_attrs to data service
- other commit that removed the service part

Quentame committed:
review: remove unused error class
was removed:
- 040cf945bb5346b6d42b3782b5061a13fb7b1f6b

emontnemery reviewed (approved):
LGTM, thanks @Quentame 👍

Quentame commented:
OMG, it happened 🎉 

Original PR #81167 was oppened on october 2022 (but OK, it's not directly a home related integration).

Thanks everyone 😘
